File: backend/src/services/analysis/keywordMatcher.js
/**
 * Keyword Matcher — Exact-match scoring against a job description
 *
 * Extracts keywords from both resume and JD, computes overlap,
 * and identifies missing keywords.
 */

// Common stop words to filter out
const STOP_WORDS = new Set([
  'a', 'an', 'the', 'and', 'or', 'but', 'is', 'are', 'was', 'were', 'be', 'been',
  'being', 'have', 'has', 'had', 'do', 'does', 'did', 'will', 'would', 'could',
  'should', 'may', 'might', 'shall', 'can', 'need', 'dare', 'ought', 'used',
  'to', 'of', 'in', 'for', 'on', 'with', 'at', 'by', 'from', 'as', 'into',
  'through', 'during', 'before', 'after', 'above', 'below', 'between',
  'this', 'that', 'these', 'those', 'it', 'its', 'we', 'you', 'they', 'them',
  'our', 'your', 'their', 'my', 'his', 'her', 'who', 'which', 'what', 'where',
  'when', 'how', 'all', 'each', 'every', 'both', 'few', 'more', 'most', 'other',
  'some', 'such', 'no', 'not', 'only', 'same', 'so', 'than', 'too', 'very',
  'just', 'also', 'about', 'up', 'out', 'if', 'then', 'else', 'while',
  'including', 'must', 'able', 'etc', 'well', 'using', 'work', 'working',
  'experience', 'required', 'preferred', 'strong', 'excellent', 'good',
  'responsibilities', 'requirements', 'qualifications', 'job', 'role', 'position',
]);

/**
 * Extract meaningful keywords/phrases from text.
 * @param {string} text
 * @returns {Set<string>}
 */
export function extractKeywords(text) {
  if (!text) return new Set();

  const words = text
    .toLowerCase()
    .replace(/[^a-z0-9\s\-+#./]/g, ' ')
    .split(/\s+/)
    .filter(w => w.length > 2 && !STOP_WORDS.has(w));

  // Also extract multi-word phrases (bigrams) for technical terms
  const phrases = [];
  const wordArray = text.toLowerCase().replace(/[^a-z0-9\s\-+#.]/g, ' ').split(/\s+/).filter(Boolean);

  for (let i = 0; i < wordArray.length - 1; i++) {
    const bigram = `${wordArray[i]} ${wordArray[i + 1]}`;
    if (!STOP_WORDS.has(wordArray[i]) && !STOP_WORDS.has(wordArray[i + 1])) {
      phrases.push(bigram);
    }
  }

  return new Set([...words, ...phrases]);
}

/**
 * Score keyword match between resume and job description.
 * @param {string} resumeText
 * @param {string} jdText
 * @returns {{ score: number, matched: string[], missing: string[], total: number, details: object }}
 */
export function matchKeywords(resumeText, jdText) {
  if (!jdText) {
    return { score: 100, matched: [], missing: [], total: 0, details: {} };
  }

  const resumeKw = extractKeywords(resumeText);
  const jdKw = extractKeywords(jdText);

  // Prioritize JD keywords — which ones does the resume have?
  const matched = [];
  const missing = [];

  for (const kw of jdKw) {
    if (resumeKw.has(kw) || resumeText.toLowerCase().includes(kw)) {
      matched.push(kw);
    } else {
      missing.push(kw);
    }
  }

  // Filter missing to only meaningful/unique terms (remove duplicates of matched bigrams)
  const filteredMissing = missing.filter(m => {
    // If it's a single word that's part of a matched bigram, skip
    if (!m.includes(' ')) {
      return !matched.some(matched_kw => matched_kw.includes(m));
    }
    return true;
  });

  // Deduplicate and take top missing keywords
  const uniqueMissing = [...new Set(filteredMissing)].slice(0, 20);
  const uniqueMatched = [...new Set(matched)].slice(0, 30);

  const score = jdKw.size > 0
    ? Math.round((matched.length / jdKw.size) * 100)
    : 100;

  return {
    score: Math.min(score, 100),
    matched: uniqueMatched,
    missing: uniqueMissing,
    total: jdKw.size,
    matchRate: `${matched.length}/${jdKw.size}`,
  };
}

File: backend/app/services/analysis/keyword_matcher.py
"""
ResumeIQ — Keyword Matcher & Technical Skill Extraction
Lexical & Skill matching against Job Descriptions with boundary checks and canonical alias expansion.
"""

import re
from typing import Dict, Any, Set, List

SKILL_ALIASES = {
    "js": "javascript",
    "ts": "typescript",
    "k8s": "kubernetes",
    "kube": "kubernetes",
    "reactjs": "react",
    "react.js": "react",
    "nodejs": "node.js",
    "node": "node.js",
    "vuejs": "vue",
    "angularjs": "angular",
    "postgres": "postgresql",
    "psql": "postgresql",
    "mongo": "mongodb",
    "es6": "javascript",
    "aws": "amazon web services",
    "gcp": "google cloud platform",
    "ci/cd": "cicd",
    "ci-cd": "cicd",
    "ml": "machine learning",
    "ai": "artificial intelligence",
    "nlp": "natural language processing",
    "dl": "deep learning",
    "cv": "computer vision",
    "rest": "rest api",
    "restful": "rest api",
}

STOP_WORDS = {
    "a", "an", "the", "and", "or", "but", "is", "are", "was", "were", "be", "been",
    "being", "have", "has", "had", "do", "does", "did", "will", "would", "could",
    "should", "may", "might", "shall", "can", "need", "dare", "ought", "used",
    "to", "of", "in", "for", "on", "with", "at", "by", "from", "as", "into",
    "through", "during", "before", "after", "above", "below", "between",
    "this", "that", "these", "those", "it", "its", "we", "you", "they", "them",
    "our", "your", "their", "my", "his", "her", "who", "which", "what", "where",
    "when", "how", "all", "each", "every", "both", "few", "more", "most", "other",
    "some", "such", "no", "not", "only", "same", "so", "than", "too", "very",
    "just", "also", "about", "up", "out", "if", "then", "else", "while",
    "including", "must", "able", "etc", "well", "using", "work", "working",
    "experience", "required", "preferred", "strong", "excellent", "good",
    "responsibilities", "requirements", "qualifications", "job", "role", "position",
    "looking", "candidate", "team", "company", "opportunity", "years", "plus",
    "skills", "ability", "degree", "computer", "science", "engineering",
}

def extract_keywords(text: str) -> Set[str]:
    """Extract meaningful unigram and bigram keywords from text."""
    if not text:
        return set()

    # Replace slashes and brackets with spaces to break compound terms like fastapi/django
    cleaned = re.sub(r"[/\\(){}\[\]|,;:\n\r\t]", " ", text.lower())
    words = []
    for token in cleaned.split():
        token = token.strip(" \t\n\r.,!?:;\"'()[]{}")
        if len(token) >= 2 and token not in STOP_WORDS and not token.isdigit():
            words.append(token)
    return set(words)

def text_contains_keyword(text: str, keyword: str) -> bool:
    """Check if keyword or any alias occurs with strict word boundaries in text."""
    normalized_text = f" {text.lower()} "
    kw = keyword.lower()

    # Direct word boundary pattern
    pattern = re.compile(rf"(^|[^a-z0-9]){re.escape(kw)}([^a-z0-9]|$)", re.IGNORECASE)
    if pattern.search(normalized_text):
        return True

    # Alias matching
    for alias, canonical in SKILL_ALIASES.items():
        if canonical == kw or alias == kw:
            target_term = alias if canonical == kw else canonical
            alias_pattern = re.compile(rf"(^|[^a-z0-9]){re.escape(target_term)}([^a-z0-9]|$)", re.IGNORECASE)
            if alias_pattern.search(normalized_text):
                return True

    return False

def match_keywords(resume_text: str, jd_text: str = None) -> Dict[str, Any]:
    """Match keywords between resume and job description."""
    if not jd_text or not resume_text:
        return {
            "score": 100,
            "matched": [],
            "missing": [],
            "total": 0,
            "matchRate": "N/A",
            "type": "lexical_and_aliases",
        }

    jd_keywords = extract_keywords(jd_text)
    if not jd_keywords:
        return {
            "score": 100,
            "matched": [],
            "missing": [],
            "total": 0,
            "matchRate": "N/A",
            "type": "lexical_and_aliases",
        }

    matched = []
    missing = []

    for kw in jd_keywords:
        if text_contains_keyword(resume_text, kw):
            matched.append(kw)
        else:
            missing.append(kw)

    unique_matched = list(dict.fromkeys(matched))[:35]
    unique_missing = list(dict.fromkeys(missing))[:25]

    score = round((len(matched) / len(jd_keywords)) * 100) if jd_keywords else 100

    return {
        "score": min(score, 100),
        "matched": unique_matched,
        "missing": unique_missing,
        "total": len(jd_keywords),
        "matchRate": f"{len(matched)}/{len(jd_keywords)}",
        "type": "lexical_and_aliases",
    }
